### PR TITLE
Show state delta in steps

### PR DESCRIPTION
**Pull request by acidjunk**
Todo:

- [x] move state delta function to utils and try to remove the only `ts-ignore` in it
- [x] add a test for state delta function
- [x] translations
- [x] implement GUI stuff for failed steps
- [x] implement GUI stuff for input steps
- [x] clean up remaining inline style definitions in `WFOSteps` (and maybe also in `WFOStepList`?)

### CODE REVIEW

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 1]
```diff
@@ -11,6 +11,33 @@ export interface WFOStepListProps {
     startedAt: string;
 }
 
+export const stateDelta = (prev: State, curr: State) => {
```
Comment (DutchBen) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:14]: Matter of taste but I'm more for complete words instead of abbrv. So previous instead of prev

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 2]
```diff
@@ -104,6 +131,32 @@ export const WFOStepList = ({ steps = [], startedAt }: WFOStepListProps) => {
             </EuiFlexGroup>
             <>
                 {steps.map((step, index) => {
+                    let prevState = {};
+                    let delta = {};
+
+                    if (index > 0) {
+                        // Todo: decided if we want this here.
+                        //  V1 has a rather big case statement with businness logic here
+                        if (steps[index].status === 'success') {
```
Comment (DutchBen) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:140]: There is a enum with statusses somewhere

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 3]
```diff
@@ -104,6 +131,32 @@ export const WFOStepList = ({ steps = [], startedAt }: WFOStepListProps) => {
...
+                            // Prepare a delta with the last successful step
+                            let prev_index = index - 1;
```
Comment (DutchBen) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:142]: camelCase or snake_case? I see both used. I prefer camelCase but at least it should be consistent

**Comment by DutchBen** [comment]
General comment. Is see there are todo's in the pr, do we still do a review if there are open todo's? What do you guys prefer?

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx [hunk 4]
```diff
@@ -89,6 +98,16 @@ export const WFOStep = ({
...
                 </EuiFlexGroup>
             </EuiFlexGroup>
+            {stepDetailIsOpen && (
+                <EuiCodeBlock
+                    isCopyable={true}
+                    language={'json'}
+                    lineNumbers={true}
+                    overflowHeight={6000}
+                >
+                    {JSON.stringify(delta, null, 3)}
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:108]: Maybe it is me, but I cannot stand a spacing of 3. Normally IDE's have either 2 or 4 configured as well. Personally I am team "4-spaces", but can live with 2 spaces as well.

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx [hunk 5]
```diff
@@ -37,6 +45,7 @@ export const WFOStep = ({
                 css={{
                     gap: 0,
                     alignItems: 'center',
+                    marginBottom: 20,
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:48]: Try to keep styling in a separate styles.ts file. More important, try to use the base value (theme.base) -- see packages/orchestrator-ui-components/src/components/WFOTimeline/styles.ts as example

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx (hunk 2 above)
Comment (ricardovdheijden) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:140]: `step[index]` is the same as `step` in this case

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 6]
```diff
@@ -104,6 +131,32 @@ export const WFOStepList = ({ steps = [], startedAt }: WFOStepListProps) => {
             </EuiFlexGroup>
             <>
                 {steps.map((step, index) => {
+                    let prevState = {};
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:134]: For me personally it is very hard to follow what is going on in this mapper, especially with the reassignments of prevState and delta. Try to move this logic as much as possible into a separate function, perferably in a separate ts file with unit tests.

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 7]
```diff
@@ -104,6 +131,32 @@ export const WFOStepList = ({ steps = [], startedAt }: WFOStepListProps) => {
...
+
+                    if (index > 0) {
+                        // Todo: decided if we want this here.
+                        //  V1 has a rather big case statement with businness logic here
+                        if (steps[index].status === 'success') {
+                            // Prepare a delta with the last successful step
+                            let prev_index = index - 1;
+                            while (
+                                prev_index > 0 &&
+                                (steps[prev_index].status === 'failed' ||
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:145]: To make this function a bit more pure: you could use the 3rd argument from the mapper function that is the whole array too

File: packages/orchestrator-ui-components/src/types/types.ts [hunk 8]
```diff
@@ -225,12 +225,16 @@ export enum ProcessStatus {
     COMPLETED = 'completed',
 }
 
+export interface State {
+    [index: string]: object;
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/types/types.ts:229]: According to data on our test environment the type of the props can also be string, boolean, null (and maybe more)
image

File: packages/orchestrator-ui-components/src/types/types.ts [hunk 9]
```diff
@@ -225,12 +225,16 @@ export enum ProcessStatus {
     COMPLETED = 'completed',
 }
 
+export interface State {
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/types/types.ts:228]: Suggesting to name the type StepState or something more descriptive

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx [hunk 10]
```diff
@@ -11,6 +11,33 @@ export interface WFOStepListProps {
...
+
+    const newState = newKeys.sort().reduce((acc: State, key) => {
+        if (
+            curr[key] === Object(curr[key]) &&
+            !Array.isArray(curr[key]) &&
+            prevOrEmpty[key]
+        ) {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            acc[key] = stateDelta(prevOrEmpty[key], curr[key]);
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:32]: Could you specify -- add as code comment -- why the TS ignore is needed here?

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx (hunk 1 above)
Comment (ricardovdheijden) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:14]: Maybe even previousState (or if agree with my type name suggestion in a different comment previousStepState)

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx (hunk 5 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:48]: So why is the gap and alignItems ok here?

File: packages/orchestrator-ui-components/src/types/types.ts (hunk 8 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/types/types.ts:229]: Yes; I wonder why the step state was defined in a way that it's incompatible with the graphql response in the first place...

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx (hunk 3 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:142]: A mix was used in v1: and this was a copy, not yet ready for reviewing.

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx (hunk 10 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStepList/WFOStepList.tsx:32]: No; it's was already in the todo. I will when it's still needed after the mentioned refactor.

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx (hunk 5 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:48]: There are at least 6 other occurences of inline CSS. I will resolve them; no problem -> but you might review stuff better before it's merged. Added it to the todo list.

Comment (ricardovdheijden) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:48]: Only since recent I came up with a pattern to move the inline css out of the components and having the theme available in that separate file. You will find many more places in the codebase where we still have inline CSS. So my comment about inline you can see as a "you could....". Though I would like to suggest to use the base value from the theme as a base to come up with these values

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx (hunk 4 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx:96]: 4 won't fit with large workflows (e.g. fw)
2 looks flimsy

File: packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx (hunk 5 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/components/WFOSteps/WFOStep/WFOStep.tsx]: @OFFTOPIC:
I now see a lot of css in getStyles.ts which hard codes color. This will get hard to maintain with a dark theme.
We need to come up with some good rules for colors.

File: packages/orchestrator-ui-components/src/types/types.ts (hunk 8 above)
Comment (acidjunk) [reply packages/orchestrator-ui-components/src/types/types.ts]: If you know a better solution please apply it in the MVP branch

**Comment by DutchBen** [comment]
Already incorporated in MVP Branch
